provisioner/aws-manager.js

```diff
@@ -166,6 +271,254 @@ AwsManager.prototype._filterSpotRequests = function (spotReqs) {
+ */
+AwsManager.prototype._reconcileStateDifferences = function (differences, deadState, apiState) {
+  var that = this;
+  assert(differences);
+  assert(deadState);
+  assert(apiState);
+
+  function plotSpotFulfilment (request) {
+    // Once we go from open -> active with a status of fulfilled we can log this
+    // spot request as successfully fulfilled.  This does not imply that
+    that.reportSpotRequestsFulfilled({
+      provisionerId: that.provisionerId,
+      region: request.Region,
+      az: request.LaunchSpecification.Placement.AvailabilityZone,
+      instanceType: request.LaunchSpecification.InstanceType,
+      workerType: request.WorkerType,
+      id: request.SpotInstanceRequestId,
+      instanceId: request.InstanceId,
+      time: dateForInflux(request.Status.UpdateTime),
+    });
+    debug('spot request %j fulfilled!', request);
+  }
+
+  // We want to figure out what happened to each of the spot requests which are
+  // no longer showing up as pending.  These can either be fulfilled or
+  // rejected.  FOr those which are fulfilled we want to create a data point
+  // which will let us trend how long spot request fulfilment takes.  For those
+  // which fail we want stats on why they failed so we can later analyze this
+  differences.requests.forEach(function (request) {
+    if (request.State === 'active' && request.Status.Code === 'fulfilled') {
+      plotSpotFulfilment(request);
+    } else if (request.State === 'open') {
+      // Here we have those spot requests which are no longer unfulfilled but
+      // do not have an api state which reflects that
+      that.__awaitingSpotFulfilmentStatus.push({
+        id: request.SpotInstanceRequestId,
+        time: new Date(),
+        iterationCount: 0,
+      });
+    } else {
+      // I wonder if we should be reporting spot requests which failed because
+      // of bid amount here in their own special series for use in trying to
+      // figure out what to bid.  We could use refused spot bids as a lower
+      // limit for bids
+      that.reportSpotRequestsDied({
+        provisionerId: that.provisionerId,
+        region: request.Region,
+        az: request.LaunchSpecification.Placement.AvailabilityZone,
+        instanceType: request.LaunchSpecification.InstanceType,
+        workerType: request.WorkerType,
+        id: request.SpotInstanceRequestId,
+        time: dateForInflux(request.Status.UpdateTime),
+        bid: parseFloat(request.SpotPrice, 10),
+        state: request.State,
+        statusCode: request.Status.Code,
+        statusMsg: request.Status.Message,
+      });
+      debug('spot request %j did something else', request);
+    }
+  });
+
+  this.__awaitingSpotFulfilmentStatus = this.__awaitingSpotFulfilmentStatus.filter(function (requestAwaiting) {
+    var keepInTheList = true;
+
+    deadState.requests.forEach(function (requestMightHave) {
+      if (requestMightHave.SpotInstanceRequestId === requestAwaiting) {
```

Comment:
Hmm, have you considered using `_.find(deadState.requests, r => r.SpotInstanceRequestId === requestAwaiting);`

If there is non. `_.find` will return `null`. You can dramatically reduce number of indentation.. and 
When you reach `keepInTheList = false` you can just `return false` immediately instead, example:
```js
this.__awaitingSpotFulfilmentStatus = this.__awaitingSpotFulfilmentStatus.filter(requestAwaiting => {
  // Look for spot request, and report it if furfilled
  var spotReq = _.find(deadState.requests, r => r.SpotInstanceRequestId === requestAwaiting);
  if (spotReq && spotReq.State === 'active' && spotReq.Status.Code === 'furfilled') {
    plotSpotFulfilment(spotReq);
    return false; // Remove from list, when reported.
  }

  return requestAwaiting.iterationCount++ <= MAX_ITERATIONS_FOR_STATE_RESOLUTION;
});
```